### PR TITLE
Participant privacy

### LINKED ISSUE #169: Limiting the list of associated authors
It's great that Wafer lets me add additional authors to a talk. However, this means that currently anyone submitting a talk has access to the list of registered users (and no way to specify co-authors not yet registered). While the latter is only a nuisance and requires later editing, the former has been considered a privacy problem in the DebConf context.

Would you entertain an alternative approach? This could be two-fold:

1. Provide a text-field allowing the free-form addition of additional authors. Entries here could be matched against the username of the Users table to give access to augmented information (e.g. show the full name), but could also just be displayed as-is in case there is no match. This can be abused and might yield orphan data, but the information is purely informational anyway.

2. Allow users to "claim talks", causing them to show up in the list of possible authors for a talk, from which then the author can select to confirm co-authorship. This is a bit more complex, but IMHO the right way to do things.

### PR DESCRIPTION
**Pull request by stefanor**
Fixes #169 

DebConf would prefer to not make the list of registered users public. Talk admins can set alternate speakers, where necessary.

### CODE REVIEW

File: wafer/settings.py [hunk 1]
```diff
@@ -265,6 +263,9 @@
 # For REGISTRATION_MODE == 'form', the form to present
 WAFER_REGISTRATION_FORM = 'wafer.users.forms.ExampleRegistrationForm'
 
+# Allow registered and anonymous users to see registered users
+WAFER_PUBLIC_ATTENDEE_LIST = False
```
Comment (drnlm) [inline wafer/settings.py]: Do we really want this to default to False? Keeping the list private doesn't strike me as what most people will want

**Comment by drnlm** [comment]
The user list isn't linked from anywhere on the site by default, and people can still play guessing games to access the indivual profile pages with this approach, so, other than making it harder to select additional authors, I'm struggling to see the significant gain here.

**Comment by stefanor** [comment]
> The user list isn't linked from anywhere on the site by default

Code is open, though.

> people can still play guessing games to access the individual profile pages with this approach

Good point. That's easy enough to 404, for people with unapproved talks, too

File: wafer/users/views.py [hunk 2]
```diff
@@ -87,7 +96,10 @@ class RegistrationView(EditOneselfMixin, FormView):
     template_name = 'wafer.users/registration.html'
 
     def get_user(self):
-        return UserProfile.objects.get(user__username=self.kwargs['username'])
+        try:
+            return UserProfile.objects.get(user__username=self.kwargs['username'])
+        except UserProfile.DoesNotExist:
+            raise Http404
```
Comment (hodgestar) [inline wafer/users/views.py]: Should we be doing `raise Http404("Reason")`? And in the other views above?

File: wafer/users/views.py [hunk 3]
```diff
@@ -38,6 +38,13 @@ class ProfileView(DetailView):
     slug_field = 'username'
     slug_url_kwarg = 'username'
 
+    def get_object(self, *args, **kwargs):
+        object_ = super(ProfileView, self).get_object(*args, **kwargs)
+        if not settings.WAFER_PUBLIC_ATTENDEE_LIST:
+            if not object_.userprofile.accepted_talks().exists():
+                raise Http404()
```
Comment (drnlm) [inline wafer/users/views.py]: Surely this is too restrictive? Logged in users should be able to see their own profile, even if they don't have an accepted talk.

File: wafer/talks/forms.py [hunk 4]
```diff
@@ -20,6 +20,14 @@ def get_talk_form_class():
 class TalkForm(forms.ModelForm):
     def __init__(self, *args, **kwargs):
         self.user = kwargs.pop('user')
+        initial = kwargs.setdefault('initial', {})
+        authors = initial.setdefault('authors', [self.user])
+
+        if not (settings.WAFER_PUBLIC_ATTENDEE_LIST
+                or self.user.has_perm('talks.change_talk')):
+            self.base_fields['authors'].limit_choices_to = {
+                'id__in': [author.id for author in authors]}
```
Comment (hodgestar) [inline wafer/talks/forms.py]: Just confirming that ordinary users then cannot add anyone else to their talk? I think that's okay, just checking that other agree.

**Comment by hodgestar** [comment]
Left a couple of thoughts, otherwise looks okay to me.

File: wafer/talks/forms.py (hunk 4 above)
Comment (drnlm) [inline wafer/talks/forms.py]: If we're not going to have the user list public, I don't think we have any better option than to restrict access to the author list like this.

File: wafer/users/views.py (hunk 3 above)
Comment (stefanor) [inline wafer/users/views.py]: Yeah, that's what happens when I rush a commit to get to lunch :(

File: wafer/users/views.py (hunk 2 above)
Comment (stefanor) [inline wafer/users/views.py]: The point of this is not to give any reason. If we give a reason, we have to handle the privacy case specially.

File: wafer/talks/forms.py (hunk 4 above)
Comment (stefanor) [inline wafer/talks/forms.py]: Yeah, admins will have to do it. There's no way to allow that, without having privacy problems, really.

File: wafer/users/views.py (hunk 2 above)
Comment (hodgestar) [inline wafer/users/views.py]: The reason isn't shown to anyone unless the `DEBUG=True` (at least according to the Django docs -- https://docs.djangoproject.com/en/1.9/topics/http/views/#the-http404-exception).

**Comment by hodgestar** [comment]
:+1: from me.

**Pull request merged**